src/librustc_data_structures/unify_log.rs

```diff
@@ -0,0 +1,133 @@
+use rustc_index::vec::{Idx, IndexVec};
+
+use ena::undo_log::{Rollback, UndoLogs};
+
+pub enum Undo<T> {
+    Move { index: T, old: usize },
+    Extend { group_index: usize, len: usize },
+    NewGroup { index: T },
+}
+
+/// Tracks which variables (represented by indices) that has been unified with eachother.
+/// Since there is often only a few variables that are interesting one must call `watch_variable`
+/// before this records any variables unified with that variable.
+///
+/// NOTE: The methods on this expect and only work correctly if a root variable from
+/// an `UnificationTable` is provided.
```

Comment:
Hmm, so, the ena code used to have logic that allowed us to reconstruct the set of variables things were unified with. Maybe it still does, but I think it was "feature-gated" as it did slow down overall unification performance. Does this unify log only run conditionally or on some variables? I don't have a very clear picture of that right now.

Reply:
Yes, that is watch `watch_variable` and the reference counting is for. I haven't compared what it would cost to always track the variables but that might be an idea. There seems to be an unused congruence-closure feature in ena so.

Reply:
Yes, if you enable that feature, it also enables the fields in ena that allow us to track the "unification set".

Reply:
The support have been removed though, only the feature itself remains https://github.com/rust-lang/ena/pull/6

Reply:
OK, well, I would want to follow up with that in a subsequent PR anyway I suspect, as this one has been around too long as it is =)